cloudinit/config/cc_ubuntu_advantage.py

```diff
@@ -288,35 +294,78 @@ def configure_ua(token, enable=None):
         # Allow `ua attach` to fail in already attached machines
         subp.subp(attach_cmd, rcs={0, 2}, logstring=redacted_cmd)
     except subp.ProcessExecutionError as e:
-        error = str(e).replace(token, REDACTED)
-        msg = f"Failure attaching Ubuntu Advantage:\n{error}"
+        err = str(e).replace(token, REDACTED)
+        msg = f"Failure attaching Ubuntu Advantage:\n{err}"
         util.logexc(LOG, msg)
         raise RuntimeError(msg) from e
 
-    enable_errors = []
-    for service in enable:
-        try:
-            cmd = ["ua", "enable", "--assume-yes", service]
-            subp.subp(cmd, capture=True)
-        except subp.ProcessExecutionError as e:
-            if re.search("is already enabled.", str(e)):
-                LOG.debug('Service "%s" already enabled.', service)
-            else:
-                enable_errors.append((service, e))
+    # Enable services
+    if not enable:
+        return
+    cmd = ["ua", "enable", "--assume-yes", "--format", "json", "--"] + enable
+    try:
+        enable_stdout, _ = subp.subp(cmd, capture=True, rcs={0, 1})
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Error while enabling service(s): {_format_services(enable)}"
+        ) from e
+
+    try:
+        enable_resp = json.loads(enable_stdout)
+    except json.JSONDecodeError as e:
+        raise RuntimeError(f"UA response was not json: {enable_stdout}") from e
+
+    # At this point we were able to load the json response from UA. This
+    # response contains a list of errors under the key 'errors'. E.g.
+    #
+    #   {
+    #      "errors": [
+    #        {
+    #           "message": "UA Apps: ESM is already enabled ...",
+    #           "message_code": "service-already-enabled",
+    #           "service": "esm-apps",
+    #           "type": "service"
+    #        },
+    #        {
+    #           "message": "Cannot enable unknown service 'asdf' ...",
+    #           "message_code": "invalid-service-or-failure",
+    #           "service": null,
+    #           "type": "system"
+    #        }
+    #      ]
+    #   }
+    #
+    # From our pov there are two type of errors, service and non-service
+    # related. We can distinguish them by checking if `service` is non-null
+    # or null respectively.
+
+    try:
+        from uaclient.messages import ALREADY_ENABLED  # noqa
+
+        UA_MC_ALREADY_ENABLED = ALREADY_ENABLED.name
+    except ImportError:
+        UA_MC_ALREADY_ENABLED = "service-already-enabled"
+
```

Comment:
I don't think we need/want this ImportError handling anymore.... `27.11.2` is released anywhere cloud-init will be published. So we should have this import problem.

Reply:
While we are touching this file, we can also drop the try/ImportErrors around `uaclient.api.u.pro.attach.auto.should_auto_attach.v1` too as that's also published now.
Might as well just pull these imports up out of the functions and up to the top of the cc_ubuntu_advantage module.

Reply:
Safe imports removed. I keep them within the functions to avoid import errors in instances without `uaclient` installed. For example, someone building the deb package on a debian machine would get import errors at autopkgtest-time.

Reply:
Oops. I stand corrected. Totally agree with you on that, then downstreams without ``uaclient` don't have to worry about not packaging it. +1